### PR TITLE
Add environment variables with the span, trace id and backend URLs

### PR DESCRIPTION
**Pull request by v1v**
### What

Add environment variables with the root span, trace id and the backend URLs, if any.

Variables names:
* `SPAN_ID`, for the span id
* `PARENT_ID`, for the span id
* `TRACE_ID`, for the trace id
* `TRACEPARENT`, for the span and trace id as explained in https://www.w3.org/TR/trace-context/#traceparent-header
* If any backend enabled then:
  * `OTEL_ELASTIC_URL`, for the URL with the trace id in Kibana.
  * `OTEL_JAEGER_URL`, for the URL with the trace id in Jaeger.
  * `OTEL_CUSTOM_URL`, for the URL with the trace id in the custom backend.

NOTE: `OTEL` is the official acronym for the **O**pen **Tel**emetry project (https://github.com/open-telemetry/opentelemetry-specification#acronym)

Add JCasC tests as stated in https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/PLUGINS.md

### Why

To help to access the open telemetry context in the build context, and be able to integrate with some other tools such as email/slack integrations and some other post build actions.

### Test

```groovy
pipeline {
  agent any
  stages {
    stage('log with info') {
      steps { sh 'env | sort' }
    }
  }
}
```
image

image

### CODE REVIEW

File: src/test/resources/io/jenkins/plugins/opentelemetry/jcasc/configuration-as-code-expected.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+endpoint: "otel-collector-contrib:4317"
+observabilityBackends:
+- elasticBackend:
+    kibanaBaseUrl: "http://localhost:5601"
+- jaegerBackend:
+    jaegerBaseUrl: "http://localhost:16686"
+- custom:
```
Comment (v1v) [inline src/test/resources/io/jenkins/plugins/opentelemetry/jcasc/configuration-as-code-expected.yml:7]: For some reason the JCasC exported output uses `custom` ?!

File: src/main/java/io/jenkins/plugins/opentelemetry/backend/ObservabilityBackend.java [hunk 2]
```diff
@@ -35,6 +35,9 @@
     @CheckForNull
     public abstract String getIconPath();
 
+    @CheckForNull
+    public abstract String getEnvVariable();
```
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/backend/ObservabilityBackend.java:39]: `getEnvVariableName()`

File: src/main/java/io/jenkins/plugins/opentelemetry/job/OtelEnvironmentContributor.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import hudson.model.EnvironmentContributor;
+import hudson.model.Run;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+@Extension
+public class OtelEnvironmentContributor extends EnvironmentContributor {
+
+    public static final String OTEL_SPAN_ID = "OTEL_SPAN_ID";
```
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/OtelEnvironmentContributor.java:21]: Shall we ask to the OpenTelemetry community?
Note that W3C Trace Context uses `traceparent` that is comprised of `trace-id`and `parent-id` (aka `span-id`):  https://www.w3.org/TR/trace-context/#traceparent-header
Maybe we could just use `TRACEPARENT`, `TRACE_ID`and `PARENT_ID`(or `SPAN_ID`).
Note that to reuse the trace-id and span-id in an HTTP call (curl, wget), user should `traceparent`.

Examples:

* https://docs.microsoft.com/en-us/azure/azure-monitor/app/correlation
```
curl --header "traceparent: 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01" localhost:8080
```

* https://github.com/open-telemetry/opentelemetry-python-contrib/issues/162
```
curl 'http://127.0.0.1:5000/api/graphql'
-H 'Accept: application/json, text/plain, /'
-H 'Referer: http://localhost:8080/somepage'
-H 'DNT: 1'
*-H 'traceparent: 00-a9c3b99a95cc045e573e163c3ac80a77-d99d251a8caecd06-01' *
-H 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36 Edg/85.0.564.70'
-H 'Content-Type: application/json'
```

File: src/test/resources/io/jenkins/plugins/opentelemetry/jcasc/configuration-as-code-expected.yml (hunk 1 above)
Comment (cyrille-leclerc) [reply src/test/resources/io/jenkins/plugins/opentelemetry/jcasc/configuration-as-code-expected.yml:7]: As discussed, we have a problem with `@Symbol`used in the wrong place.

Comment (v1v) [reply src/test/resources/io/jenkins/plugins/opentelemetry/jcasc/configuration-as-code-expected.yml]: https://github.com/cyrille-leclerc/opentelemetry-plugin/pull/14/commits/4184a32e70d5ea1b38b8694dd887fb2f743f4aef

**Comment by cyrille-leclerc** [comment]
I'm not exactly sure it is caused by this PR but when I build the code, I get a JCASC error:

```
$ mvn clean verify

...

[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running InjectedTest
[INFO] Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 15.297 s - in InjectedTest
[INFO] Running io.jenkins.plugins.opentelemetry.JenkinsOtelPluginIntegrationTest
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.357 s <<< FAILURE! - in io.jenkins.plugins.opentelemetry.JenkinsOtelPluginIntegrationTest
[ERROR] io.jenkins.plugins.opentelemetry.JenkinsOtelPluginIntegrationTest  Time elapsed: 1.355 s  <<< FAILURE!
java.lang.AssertionError: Unexpected exception 
        at io.jenkins.plugins.casc.misc.JenkinsConfiguredWithCodeRule.before(JenkinsConfiguredWithCodeRule.java:50)
        at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:595)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: io.jenkins.plugins.casc.ConfiguratorException: unclassified: error configuring 'unclassified' with class io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator configurator
        at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:697)
        at io.jenkins.plugins.casc.ConfigurationAsCode.checkWith(ConfigurationAsCode.java:755)
        at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:740)
        at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:616)
        at io.jenkins.plugins.casc.ConfigurationAsCode.configure(ConfigurationAsCode.java:585)
        at io.jenkins.plugins.casc.misc.JenkinsConfiguredWithCodeRule.before(JenkinsConfiguredWithCodeRule.java:47)
        ... 5 more
Caused by: io.jenkins.plugins.casc.ConfiguratorException: Invalid configuration elements for type class jenkins.model.GlobalConfigurationCategory$Unclassified : openTelemetry.
Available attributes : administrativeMonitorsConfiguration, artifactManager, buildDiscarders, casCGlobalConfig, defaultFolderConfiguration, defaultView, gitSCM, globalDefaultFlowDurabilityLevel, globalLibraries, junitTestResultStorage, location, mailer, masterBuild, myView, nodeProperties, plugin, pollSCM, projectNamingStrategy, quietPeriod, resourceRoot, scmRetryCount, shell, usageStatistics, viewsTabBar
        at io.jenkins.plugins.casc.BaseConfigurator.handleUnknown(BaseConfigurator.java:376)
        at io.jenkins.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:365)
        at io.jenkins.plugins.casc.BaseConfigurator.check(BaseConfigurator.java:287)
        at io.jenkins.plugins.casc.ConfigurationAsCode.lambda$checkWith$8(ConfigurationAsCode.java:755)
        at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:691)
        ... 10 more

[INFO] Running io.jenkins.plugins.opentelemetry.MonitoringActionTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.129 s - in io.jenkins.plugins.opentelemetry.MonitoringActionTest
[INFO] Running io.jenkins.plugins.opentelemetry.OpenTelemetryTest
[WARNING] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.004 s - in io.jenkins.plugins.opentelemetry.OpenTelemetryTest
[INFO] Running io.jenkins.plugins.opentelemetry.jcasc.ConfigurationAsCodeTest
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.3 s <<< FAILURE! - in io.jenkins.plugins.opentelemetry.jcasc.ConfigurationAsCodeTest
[ERROR] io.jenkins.plugins.opentelemetry.jcasc.ConfigurationAsCodeTest  Time elapsed: 2.298 s  <<< FAILURE!
java.lang.AssertionError: Unexpected exception 
        at io.jenkins.plugins.casc.misc.JenkinsConfiguredWithCodeRule.before(JenkinsConfiguredWithCodeRule.java:50)
        at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:595)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: io.jenkins.plugins.casc.ConfiguratorException: unclassified: error configuring 'unclassified' with class io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator configurator
        at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:697)
        at io.jenkins.plugins.casc.ConfigurationAsCode.checkWith(ConfigurationAsCode.java:755)
        at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:740)
        at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:616)
        at io.jenkins.plugins.casc.ConfigurationAsCode.configure(ConfigurationAsCode.java:585)
        at io.jenkins.plugins.casc.misc.JenkinsConfiguredWithCodeRule.before(JenkinsConfiguredWithCodeRule.java:47)
        ... 5 more
Caused by: io.jenkins.plugins.casc.ConfiguratorException: Invalid configuration elements for type class jenkins.model.GlobalConfigurationCategory$Unclassified : openTelemetry.
Available attributes : administrativeMonitorsConfiguration, artifactManager, buildDiscarders, casCGlobalConfig, defaultFolderConfiguration, defaultView, gitSCM, globalDefaultFlowDurabilityLevel, globalLibraries, junitTestResultStorage, location, mailer, masterBuild, myView, nodeProperties, plugin, pollSCM, projectNamingStrategy, quietPeriod, resourceRoot, scmRetryCount, shell, usageStatistics, viewsTabBar
        at io.jenkins.plugins.casc.BaseConfigurator.handleUnknown(BaseConfigurator.java:376)
        at io.jenkins.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:365)
        at io.jenkins.plugins.casc.BaseConfigurator.check(BaseConfigurator.java:287)
        at io.jenkins.plugins.casc.ConfigurationAsCode.lambda$checkWith$8(ConfigurationAsCode.java:755)
        at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:691)
        ... 10 more

[INFO] Running io.jenkins.plugins.opentelemetry.job.GitStepHandlerTest
[INFO] Tests run: 16, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.007 s - in io.jenkins.plugins.opentelemetry.job.GitStepHandlerTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   JenkinsOtelPluginIntegrationTest Unexpected exception 
[ERROR]   ConfigurationAsCodeTest Unexpected exception 
[INFO] 
[ERROR] Tests run: 29, Failures: 2, Errors: 0, Skipped: 1
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  37.776 s
[INFO] Finished at: 2021-02-15T22:59:42+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M4:test (default-test) on project opentelemetry: There are test failures.
[ERROR] 
[ERROR] Please refer to /Users/cyrilleleclerc/git/jenkins/opentelemetry/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

```

**Pull request merged**